go/plugins/weaviate/weaviate.go

```diff
@@ -0,0 +1,374 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package weaviate
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/firebase/genkit/go/ai"
+	"github.com/weaviate/weaviate-go-client/v4/weaviate"
+	"github.com/weaviate/weaviate-go-client/v4/weaviate/auth"
+	"github.com/weaviate/weaviate-go-client/v4/weaviate/graphql"
+	"github.com/weaviate/weaviate/entities/models"
+)
+
+// The provider used in the registry.
+const provider = "weaviate"
+
+// The metadata key used to hold document text.
+const textKey = "text"
+
+// The metadata key to hold document metadata.
+const metadataKey = "metadata"
+
+// state holds the current plugin state.
+var state struct {
+	mu          sync.Mutex
+	initialized bool
+	client      *weaviate.Client
+}
+
+// getClient returns the client stored in the state.
+func getClient() *weaviate.Client {
+	state.mu.Lock()
+	defer state.mu.Unlock()
+	return state.client
+}
+
+// ClientConfig passes configuration options to the plugin.
+type ClientConfig struct {
+	// The hostname:port to use to contact the Weaviate database.
+	// If not set, the default is read from the WEAVIATE_URL environment variable.
+	Addr string
+	// The scheme to use to contact the Weaviate database.
+	// Typically http or https.
+	// If not set, the default is https.
+	Scheme string
+	// The API key to use with the Weaviate database.
+	// If not set, the default is read from the WEAVIATE_API environment variable.
```

Comment:
Elsewhere we use "FOO_API_KEY".

Reply:
Done.  I think I borrowed WEAVIATE_API from a weaviate example but may as well be consistent with genkit.